python/ray/tune/schedulers/resource_changing_scheduler.py

```diff
@@ -0,0 +1,334 @@
+        requirements to values bigger than possible, the trial will
+        not run. Ensure that your function accounts for that possibility
+        by setting upper limits. Consult :func:`evenly_distribute_cpus_gpus`
+        to see how that may be done.
+
+    Example:
+        .. code-block:: python
+
+            base_scheduler = ASHAScheduler(max_t=16)
+            def my_resources_allocation_function(
+                trial_runner: "trial_runner.TrialRunner",
+                trial: Trial,
+                result: Dict[str, Any],
+                base_trial_resource: Union[PlacementGroupFactory, Resources]
+            ) -> Union[None, PlacementGroupFactory, Resource]:
+                # logic here
+                # usage of PlacementGroupFactory is strongly preferred
+                return PlacementGroupFactory(...)
+            scheduler = ResourceChangingScheduler(
+                            base_scheduler,
+                            my_resources_allocation_function
+                        )
+
+        See :doc:`/tune/examples/xgboost_dynamic_resources_example` for a
+        more detailed example.
+    """
+
+    def __init__(
+            self,
+            base_scheduler: Optional[TrialScheduler] = None,
+            resources_allocation_function: Optional[Callable[[
+                "trial_runner.TrialRunner", Trial, Dict[str, Any], Union[
+                    PlacementGroupFactory, Resources]
+            ], Union[None, PlacementGroupFactory,
+                     Resources]]] = evenly_distribute_cpus_gpus,
+    ) -> None:
+        super().__init__()
+        if resources_allocation_function is None:
+            warnings.warn(
+                "`resources_allocation_function` is None. No resource "
+                "requirements will be changed at any time. Pass a "
+                "correctly defined function to enable functionality.")
+        self._resources_allocation_function = resources_allocation_function
+        self._base_scheduler = base_scheduler or FIFOScheduler()
+        self._base_trial_resources: Optional[Union[
+            Resources, PlacementGroupFactory]] = None
+        self._trials_to_reallocate: Dict[Trial, Union[
+            None, dict, PlacementGroupFactory]] = {}
+
+    @property
+    def metric(self):
+        return self._base_scheduler._metric
+
+    def set_search_properties(self, metric: Optional[str],
+                              mode: Optional[str]) -> bool:
+        return self._base_scheduler.set_search_properties(metric, mode)
+
+    def on_trial_add(self, trial_runner: "trial_runner.TrialRunner",
+                     trial: Trial, **kwargs):
+        if self._base_trial_resources is None:
+            if trial.uses_placement_groups:
+                self._base_trial_resources = trial.placement_group_factory
+            else:
+                self._base_trial_resources = trial.resources
+        return self._base_scheduler.on_trial_add(trial_runner, trial, **kwargs)
```

Comment:
Should we enforce that this can only be used for trials that share the same base trial resources? I.e. if base trial resources is set and we add another trial with different base trial resources, then throw an RuntimeError?

Reply:
It's not possible for that to happen now, right? If we want to futureproof it, I think we should make a "base" placement group an attribute of Trial itself

Reply:
Oh it actually is - there are examples when the `num_workers` are tuned for rllib trainables, effectively changing the placement group factories for different trials

Reply:
Didn't know that, thanks! I'll make it throw an exception as you suggested and we can think of enhancing it later

Reply:
Actually, would this change mess up with resuming from a checkpoint?